src/rfc-8252-http-server.ts

```diff
@@ -92,35 +169,89 @@ export class RFC8252HTTPServer {
       // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
       url.search = new URLSearchParams(Object.entries(req.body)).toString();
     } else if (req.method !== 'GET') {
-      this.logger.emit('mongodb-oidc-plugin:oidc-callback-rejected', {
-        method: req.method,
-        hasBody,
+      reject({
+        status: 405,
+        error: 'invalid_method',
+        errorDescription: 'Invalid HTTP Method',
       });
-      res.status(405);
-      res.type('text/plain');
-      res.send('405 Invalid Method');
       return;
     }
+
+    // If `state` does not match the expected value, this might be a CSRF request.
+    // Even if we did progress from here, the OIDC library would reject this
+    // request, but bailing out early here means that after this conditional
+    // we know we are dealing with a genuine response from the auth server and
+    // can reject the `oidcParamsPromise` if there is an error.
+    if (url.searchParams.get('state') !== this.oidcStateParam) {
+      reject({
+        status: 403,
+        error: 'state_mismatch',
+        errorDescription: '',
+      });
+      return;
+    }
+    isAcceptedOIDCResponse = true;
+
+    const oidcParams = url.toString();
+    if (this.oidcParams !== undefined && this.oidcParams !== oidcParams) {
+      // This should not happen in practice...
+      reject({
+        status: 409,
+        error: 'already_received_code',
+        errorDescription:
+          'A different authentication code was already received by the application',
+      });
+      return;
+    }
+    this.oidcParams = oidcParams;
     this.logger.emit('mongodb-oidc-plugin:oidc-callback-accepted', {
       method: req.method,
       hasBody,
+      errorCode: url.searchParams.get('error') ?? undefined,
     });
-    res.status(200);
-    res.type('text/plain');
-    res.send('Welcome! Everything is fine.');
-    this.oidcParamsResolve?.(url.toString());
+
+    if (url.searchParams.get('error')) {
+      reject({
+        status: 200,
+        // Standard params from https://www.rfc-editor.org/rfc/rfc6749.html#section-4.1.2.1
+        error:
+          ensureNQSChar(url.searchParams.get('error')) ?? 'invalid_error_param',
+        errorDescription: ensureNQSChar(
+          url.searchParams.get('error_description')
+        ),
+        errorURI: ensureURI(url.searchParams.get('error_uri')),
+      });
+      return;
+    }
+
+    // https://mailarchive.ietf.org/arch/msg/oauth/RqlUvseG_RnOWrEV_WJACW8oUdU/
+    // > Callback URL pages SHOULD redirect to a trusted page immediately after
+    // > receiving the authorization code in the URL.  This prevents the
+    // > authorization code from remaining in the browser history, or from
+    // > inadvertently leaking in a referer header.
```

Comment:
```suggestion
    // > inadvertently leaking in a referrer header.
```

Reply:
One of the most famous typos in the world :) https://stackoverflow.com/questions/3087626/was-the-misspelling-of-the-http-field-name-referer-intentional

I'd probably not fix this either way since it's a direct quote from the mailing list entry linked above.